File: backend/src/entities/AIConversationMessage.ts
import { Entity, PrimaryGeneratedColumn, Column } from "typeorm";

export type AIMessageRole = "user" | "assistant" | "system" | "tool";

/**
 * One turn in an `AIConversation`. `role: "tool"` rows record a tool's
 * result being fed back to FabriXAI (see `aiAssistantService.ts` step 6) —
 * kept for audit/debugging, never replayed as conversation history verbatim
 * to the user-facing transcript.
 */
@Entity("AIConversationMessages")
export class AIConversationMessage {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column({ type: "uniqueidentifier" })
  conversationId!: string;

  @Column({ type: "nvarchar", length: 20 })
  role!: AIMessageRole;

  @Column({ type: "nvarchar", length: "MAX" })
  message!: string;

  @Column({ type: "int", nullable: true })
  tokenUsage!: number | null;

  @Column({ type: "nvarchar", length: 100, nullable: true })
  model!: string | null;

  @Column({ type: "nvarchar", length: 100, nullable: true })
  requestId!: string | null;

  @Column({ type: "datetimeoffset", default: () => "SYSDATETIMEOFFSET()" })
  createdAt!: Date;
}

File: backend-py/app/models/ai_conversation_message.py
"""Port of `backend/src/entities/AIConversationMessage.ts`."""

from __future__ import annotations

from typing import Literal, Optional

from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, datetimeoffset_now, nvarchar, nvarchar_max, uuid_col, uuid_pk

AIMessageRole = Literal["user", "assistant", "system", "tool"]


class AIConversationMessage(Base):
    __tablename__ = "AIConversationMessages"

    id: Mapped[str] = uuid_pk()
    conversationId: Mapped[str] = uuid_col()
    role: Mapped[AIMessageRole] = nvarchar(20)
    message: Mapped[str] = nvarchar_max()
    tokenUsage: Mapped[Optional[int]] = mapped_column(Integer, nullable=True, default=None)
    model: Mapped[Optional[str]] = nvarchar(100, nullable=True)
    requestId: Mapped[Optional[str]] = nvarchar(100, nullable=True)
    createdAt = datetimeoffset_now()
